src/messaging/ExchangeContext.h

```diff
@@ -0,0 +1,271 @@
+     *                                                       CHIP message protocol version that is supported. For example,
+     *                                                       this error would be generated if CHIP Reliable Messaging
+     *                                                       semantics are being attempted when the CHIP message protocol
+     *                                                       version is V1.
+     *  @retval  #CHIP_ERROR_NOT_CONNECTED                  if the context was associated with a connection that is now
+     *                                                       closed.
+     *  @retval  #CHIP_ERROR_INCORRECT_STATE                if the state of the exchange context is incorrect.
+     *  @retval  #CHIP_NO_ERROR                             if the CHIP layer successfully sent the message down to the
+     *                                                       network layer.
+     */
+    CHIP_ERROR SendMessage(uint32_t protocolId, uint8_t msgType, System::PacketBuffer * msgPayload, uint16_t sendFlags = 0,
+                           void * msgCtxt = nullptr);
+
+    /**
+     *  Tear down the Trickle retransmission mechanism by canceling the periodic timers
+     *  within Trickle and freeing the message buffer holding the Weave
+     *  message.
+     */
+    void TeardownTrickleRetransmit();
+
+    /**
+     *  Cancel the Trickle retransmission mechanism.
+     */
+    void CancelRetrans();
+
+    /**
+     *  Set delegate of the current ExchangeContext instance.
+     */
+    void SetDelegate(ExchangeContextDelegate * delegate) { mDelegate = delegate; }
+
+    /**
+     *  Return the delegate pointer of the current ExchangeContext instance.
+     */
+    ExchangeContextDelegate * GetDelegate() { return mDelegate; }
+
+    /*
+     * In order to use reference counting (see refCount below)
+     * we use a hold/free paradigm where users of the exchange
+     * can hold onto it while it's out of their direct control
+     * to make sure it isn't closed before everyone's ready.
+     */
+    void AddRef();
+    void Close();
+    void Abort();
+    void Release();
+    CHIP_ERROR StartTimerT();
+
+private:
+    System::PacketBuffer * msg; // If we are re-transmitting, then this is the pointer to the message being retransmitted
+    ExchangeContextDelegate * mDelegate = nullptr;
+
+    // Trickle-controlled retransmissions:
+    uint32_t backoff;                        // backoff for sampling the numner of messages
+    uint8_t msgsReceived;                    // number of messages heard during the backoff period
+    uint8_t rebroadcastThreshold;            // re-broadcast threshold
+    BitFlags<uint16_t, ExFlagValues> mFlags; // Internal state flags
+
+    CHIP_ERROR ResendMessage();
+    bool MatchExchange(const PacketHeader & packetHeader, const PayloadHeader & payloadHeader);
```

Comment:
What odes this do? Is it 'same exchange'? if so, why not just an equality test on the exchange ID?

Reply:
It search for an existing exchange that the incoming message applies to.